dbms/src/Flash/Mpp/GRPCSendQueue.h

```diff
@@ -0,0 +1,264 @@
+/// a timeout and it uses a timer to trigger the notification, which is wasteful if we want
+/// to trigger it immediately. So we can say `kickCompletionQueue` function is a
+/// immediately-triggered `Alarm`.
+template <typename T>
+class GRPCSendQueue
+{
+public:
+    GRPCSendQueue(size_t queue_size, grpc_call * call, const LoggerPtr & l)
+        : send_queue(queue_size)
+        , log(l)
+        , kick_tag([this]() { return kickTagAction(); })
+    {
+        RUNTIME_ASSERT(call != nullptr, log, "call is null");
+        // If a call to `grpc_call_start_batch` with an empty batch returns
+        // `GRPC_CALL_OK`, the tag is pushed into the completion queue immediately.
+        // This behavior is well-defined. See https://github.com/grpc/grpc/issues/16357.
+        kick_func = [call](void * t) {
+            return grpc_call_start_batch(call, nullptr, 0, t, nullptr);
+        };
+    }
+
+    // For gtest usage.
+    GRPCSendQueue(size_t queue_size, GRPCKickFunc func)
+        : send_queue(queue_size)
+        , log(Logger::get("GRPCSendQueue", "test"))
+        , kick_func(func)
+        , kick_tag([this]() { return kickTagAction(); })
+    {}
+
+    ~GRPCSendQueue()
+    {
+        std::unique_lock lock(mu);
+
+        RUNTIME_ASSERT(status == Status::NONE, log, "status {} is not none", status);
+    }
+
+    /// Push the data from queue and kick the grpc completion queue.
+    ///
+    /// Return true if push succeed.
+    /// Else return false.
+    template <typename U>
+    bool push(U && u)
+    {
+        auto ret = send_queue.push(std::forward<U>(u)) == MPMCQueueResult::OK;
+        if (ret)
+        {
+            kickCompletionQueue();
+        }
+        return ret;
+    }
+
+    /// Pop the data from queue.
+    ///
+    /// Return OK if pop is done.
+    /// Return FINISHED if the queue is finished and empty.
+    /// Return EMPTY if there is no data in queue and `new_tag` is saved.
+    /// When the next push/finish is called, the `new_tag` will be pushed
+    /// into grpc completion queue.
+    /// Note that any data in `new_tag` mustn't be touched if this function
+    /// returns EMPTY because this `new_tag` may be popped out in another
+    /// grpc thread immediately. By the way, if this completion queue is only
+    /// tied to one grpc thread, this data race will not happen.
+    GRPCSendQueueRes pop(T & data, void * new_tag)
+    {
+        RUNTIME_ASSERT(new_tag != nullptr, log, "new_tag is nullptr");
+
+        auto res = send_queue.tryPop(data);
+        switch (res)
+        {
+        case MPMCQueueResult::OK:
+            return GRPCSendQueueRes::OK;
+        case MPMCQueueResult::FINISHED:
+            return GRPCSendQueueRes::FINISHED;
+        case MPMCQueueResult::EMPTY:
+            // Handle this case later.
+            break;
+        default:
+            RUNTIME_ASSERT(false, log, "Result {} is invalid", res);
+        }
+
+        std::unique_lock lock(mu);
```

Comment:
can `pop(T & data, void * new_tag)`  be called concurrently? What will happen when multiple threads call `send_queue.tryPop` and  return `MPMCQueueResult::EMPTY` before and only one thread call the following `MPMCQueueResult::EMPTY` branch and other theads found `status` is `Status::WAITING` which cause `RUNTIME_ASSERT` fail

Reply:
No, it can't. The comments above `GRPCSendQueue` says it's a `multi-producer-single-consumer` queue.